Onboarding Improvements: Epilogue screen create new site

momo-ozawa opened the pull request:
Part of #17385

## Description

This PR hooks up the **Create a new site** button on the Epilogue screen.

## Notes

- Please note that there is a [known issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/17453) where the wrong cells get displayed in the Login Epilogue screen when an account has exactly 4 sites. This will be addressed in a separate PR. 
- There are some obsolete files we should be able to delete, once this PR gets approved. This will be addressed in a separate PR. 
- UI tests will be addressed in a different PR.

## Merge directions
- Merge after #17435

## How to test

### [NEW] Epilogue -> Create a new site -> Show me around

1. Log out, then log in
2. On the Epilogue screen, tap **Create a new site**
3. Choose a design and tap **Choose**
4. Choose a domain and tap **Create Site**
5. On the Site Created success screen, tap **Done**
6. ✅ The Quick Start prompt is displayed
7. On the Quick Start prompt, tap **Show me around**
8. ✅ The Quick Start checklist is displayed
9. On the Quick Start checklist, tap the **X button**
10. ✅ The checklist is dismissed, and the Quick Start menu items (i.e. Customize Your Site / Grow Your Audience) are showing in My Site

<details>
 <summary>Click here for iPhone video</summary>

https://user-images.githubusercontent.com/6711616/141370458-e3969557-3740-4b0c-8795-0c80ee785481.mp4

</details>

<details>
 <summary>Click here for iPad video</summary>

https://user-images.githubusercontent.com/6711616/141370494-f4362d81-1a6b-47ee-8596-19aad565e1ce.mp4

</details>

### [NEW] Epilogue > Create a new site > No thanks

1. Log out, then log in
2. On the Epilogue screen, tap **Create a new site**
3. Choose a design and tap **Choose**
4. Choose a domain and tap **Create Site**
5. On the Site Created success screen, tap **Done**
6. ✅ The Quick Start prompt is displayed
7. On the Quick Start prompt, tap **No thanks**
8. ✅ The prompt is dismissed, and the Quick Start menu items (i.e. Customize Your Site / Grow Your Audience) are NOT showing in My Site

<details>
 <summary>Click here for iPhone video</summary>

https://user-images.githubusercontent.com/6711616/141370526-06986785-b9a4-48d1-b2f1-2d60b9be319e.mp4

</details>

<details>
 <summary>Click here for iPad video</summary>

https://user-images.githubusercontent.com/6711616/141370556-a6845145-64e0-4e12-8d2c-e802d2221944.mp4

</details>


### [Regression] My Site > Create a new site > Show me around

1. Log in if not logged in
2. On My Site, tap the **chevron icon** by the site title header
3. On the Site Switcher screen, tap the **+ icon**
4. Tap **Create WordPress.com site** and go through the site creation flow
5. On the Site Created success screen, tap **Done**
6. ✅ The Quick Start prompt is displayed
7. On the Quick Start prompt, tap **Show me around**
8. ✅ The Quick Start checklist is displayed
9. On the Quick Start checklist, tap the **X button**
10. ✅ The checklist is dismissed, and the Quick Start menu items (i.e. Customize Your Site / Grow Your Audience) are showing in My Site

### [Regression] My Site > Create a new site > No thanks

1. Log in if not logged in
2. On My Site, tap the **chevron icon** by the site title header
3. On the Site Switcher screen, tap the **+ icon**
4. Tap **Create WordPress.com site** and go through the site creation flow
5. On the Site Created success screen, tap **Done**
6. ✅ The Quick Start prompt is displayed
7. On the Quick Start prompt, tap **No thanks**
8. ✅ The prompt is dismissed, and the Quick Start menu items (i.e. Customize Your Site / Grow Your Audience) are NOT showing in My Site

## Regression Notes
1. Potential unintended areas of impact
Site creation > Quick Start flow

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested the above steps manually

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add: onCreateNewSite closure

Commit:
Add: present site creation wizard from epilogue

Commit:
Refactor: pass onDismissQuickStart closure

Commit:
Refactor: dismiss the fullscreen signin from the quick start prompt

Commit:
Refactor: show quick start prompt instead of old quick start alert

momo-ozawa commented:
Hi @ScoutHarris and @frosty 👋
Asking for a review from either of you since Leandro is AFK for the week, and I'm the only iOS dev on team Raven.
Lmk if you have any q's about this feature. Thanks in advance! 🙇‍♀️

@osullivanchris 
I added some videos up above in the **How to test** section, for both iPhone and iPad. Lmk what you think. Also, the installable build can be downloaded from [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/17449#issuecomment-966096546) once it's ready 😄

frosty commented on WordPress/Classes/ViewRelated/Blog/QuickStartPromptViewController.swift:
```diff
@@ -21,9 +21,9 @@ final class QuickStartPromptViewController: UIViewController {
+    /// Closure to be executed upon dismissing the Login Epilogue.
     ///
-    var onDismiss: ((Blog) -> Void)?
+    var onDismissEpilogue: (() -> Void)?
```
I feel like `onDismiss` might be a better name here, as this screen doesn't really have anything to do specifically with the login epilogue and could in theory be used in different places too.

frosty commented on WordPress/Classes/ViewRelated/Blog/QuickStartPromptViewController.swift:
```diff
@@ -101,25 +101,34 @@ final class QuickStartPromptViewController: UIViewController {
+            onDismissEpilogue()
+
+            // Show the My Site screen for the specified blog
+            WordPressAppDelegate.shared?.windowManager.dismissFullscreenSignIn(blogToShow: blog, completion: {
```
This code is perhaps making assumptions about how this screen is being displayed and that full screen sign-in is being displayed. I wonder if we would be better to remove this login specific code from this VC and make it part of the onDismiss closure itself.

So in `WordPressAuthenticationManager.swift` we'd have something like:

```
let onDismiss = { showQuickStart in
    // Show the My Site screen for the specified blog
    WordPressAppDelegate.shared?.windowManager.dismissFullscreenSignIn(blogToShow: blog, completion: {
        guard showQuickStart else {
            return
        }

        // After a short delay, trigger the Quick Start tour
        DispatchQueue.main.asyncAfter(deadline: .now() + Constants.quickStartDelay) {
            QuickStartTourGuide.shared.setup(for: self.blog)
        }
    })
}

...

quickstartPrompt.onDismissEpilogue = onDismiss
```

That way, the quick start prompt doesn't need to know about the context in which it's displayed, or the complexities of how it should dismiss itself. What do you think?

momo-ozawa replied:
Yes, 100%. That's a very good point, will fix this.

frosty commented on WordPress/Classes/ViewRelated/Site Creation/FinalAssembly/SiteAssemblyStep.swift:
```diff
@@ -25,9 +25,9 @@ final class SiteAssemblyStep: WizardStep {
     ///   - creator: the in-flight creation instance
     ///   - service: the service to use for initiating site creation
-    init(creator: SiteCreator, service: SiteAssemblyService) {
+    init(creator: SiteCreator, service: SiteAssemblyService, onDismissEpilogue: (() -> Void)? = nil) {
```
Same as with the quick start prompt, I'm not sure that this needs to reference the epilogue specifically, as the site assembly flow can be displayed at other times too.

Also, we should add this parameter to the parameters doc comment above.

frosty commented on WordPress/Classes/ViewRelated/Site Creation/FinalAssembly/SiteAssemblyWizardContent.swift:
```diff
@@ -32,16 +32,20 @@ final class SiteAssemblyWizardContent: UIViewController {
     ///   - creator: the in-flight creation instance
     ///   - service: the service to use for initiating site creation
-    init(creator: SiteCreator, service: SiteAssemblyService) {
+    init(creator: SiteCreator, service: SiteAssemblyService, onDismissEpilogue: (() -> Void)? = nil) {
```
Same comment as SiteAssemblyStep around naming and comment.

frosty reviewed (commented):
Hello! I've run through all the testing steps and it's working well for me. I checked iPhone and iPad, all good.

I've added a few code level comments around naming and moving a few things around.

Finally, one UX query from me (mainly applies to iPhone) perhaps for @osullivanchris: the positioning of the buttons on the quick start prompt screen doesn't feel great on iPhone as they're so far up the screen and away from the reachable area when holding the phone one handed:

<img src="https://user-images.githubusercontent.com/4780/141460533-2b978c21-c7a4-4ba3-9a13-9cb6534f5128.png" width="350">

Is there any way we can move them to the bottom of the screen, or display this as a smaller bottom sheet or something? Could be a future improvement rather than blocking this PR.

osullivanchris reviewed (approved):
Hey @momo-ozawa everything works as expected and looks solid.

> the positioning of the buttons on the quick start prompt screen doesn't feel great on iPhone as they're so far up the screen and away from the reachable area when holding the phone one handed

good question from @frosty . Can't 100% recall why I went with this position on Android, and kept same for iOS. I think it was because the content area is quite small, buttons fixed to the bottom might feel far away and disconnected from the question being posed to the user. But I can see in my hand that I'm tapping bottom buttons a couple times in this flow, then jumping to the top. What do you think @momo-ozawa ?

A second one. We decided this week to move the prompt to a modal presentation in this flow. Seeing it now, I feel like we need to nudge the content down a bit further. I think there should almost be the space reserved for a native navigation bar, even though we are not using one in this instance. Then the content might feel more correctly positioned. Here its almost taking the place a nav bar would, but its not one, and just looks slightly off I think.

Commit:
Merge branch 'epic/onboarding-improvements' into task/onboarding-improvements-epilogue-create-new-site-cta

# Conflicts:
#	WordPress/Classes/ViewRelated/Blog/QuickStartPromptViewController.swift
#	WordPress/Classes/ViewRelated/NUX/WordPressAuthenticationManager.swift

Commit:
Refactor: define on dismiss closure in wp auth manager

The quick start prompt shouldn't need to know about the context in which it's displayed, or the complexities of how it should dismiss itself.

momo-ozawa commented:
@frosty Thanks for reviewing! Ready for another look, addressed your feedback in 606fe63

momo-ozawa commented:
@osullivanchris 

> But I can see in my hand that I'm tapping bottom buttons a couple times in this flow, then jumping to the top. What do you think @momo-ozawa ?

I agree w you and @frosty, I will open a new PR where I move the buttons to the bottom of the screen. 

> Seeing it now, I feel like we need to nudge the content down a bit further. I think there should almost be the space reserved for a native navigation bar, even though we are not using one in this instance. Then the content might feel more correctly positioned. Here its almost taking the place a nav bar would, but its not one, and just looks slightly off I think.

I can also bump down the content and see how that looks 👍

frosty commented on WordPress/Classes/ViewRelated/NUX/WordPressAuthenticationManager.swift:
```diff
@@ -336,27 +336,45 @@ extension WordPressAuthenticationManager: WordPressAuthenticatorDelegate {
                 return
             }
 
+            // Otherwise, whow the My Site screen for the specified blog and after a short delay,
```
Nitpick, typo: `whow`

momo-ozawa replied:
🙈 Fixed! ff7e573

frosty reviewed (commented):
Code changes look good to me (one small typo!) but in my testing Quick Start never gets activated if I start Site Creation via the + in the site picker list. I think this is because `launchSiteCreation` in both MySiteViewController and BlogListViewController don't have an onDismiss closure passed to them?

Commit:
Fix: typo

Commit:
Refactor: inject quick start settings

Commit:
Refactor: extract delayed quick start execution to a method

Commit:
Fix: show quick start tour after site creation via site picker

momo-ozawa commented:
> in my testing Quick Start never gets activated if I start Site Creation via the + in the site picker list. 

@frosty Great catch, thank you! Fixed in e1b60ad.

https://user-images.githubusercontent.com/6711616/142399113-6c52e830-4299-416f-ac2d-21e18a5efbed.mp4

frosty commented on WordPress/Classes/ViewRelated/Site Creation/FinalAssembly/SiteAssemblyWizardContent.swift:
```diff
@@ -42,10 +45,15 @@ final class SiteAssemblyWizardContent: UIViewController {
+         onDismiss: ((Blog) -> Void)? = nil) {
         self.siteCreator = creator
         self.service = service
+        self.quickStartSettings = quickStartSettings ?? QuickStartSettings()
```
The init parameter could be non-optional with a default value of `QuickStartSettings()`, just to prevent the need for this :)

momo-ozawa replied:
Fixed! 72a664b

frosty commented on WordPress/Classes/ViewRelated/Blog/QuickStartTourGuide.swift:
```diff
@@ -28,6 +28,18 @@ open class QuickStartTourGuide: NSObject {
         }
     }
 
+    func setup(for blog: Blog, withCompletedSteps steps: [QuickStartTour] = [], executeWithDelay: Bool = false) {
```
Because this method is currently only ever used to add a delay (and without one it does the same as `setup(for:withCompletedSteps:)` I guess it could be renamed `delayedSetup(for:withCompletedSteps:)` or `setupWithDelay(for:withCompletedSteps:)` and then you could remove the guard entirely?

momo-ozawa replied:
Fixed! 2432d7d

frosty reviewed (approved):
Changes look good! Thank you for putting up with all my suggestions 🙂 

I just have two more small ones, but up to you if you want to implement them or not. Otherwise, :shipit:!

Commit:
Refactor: rename method to setupWithDelay and omit executeWithDelay param

Commit:
Refactor: make QuickStartSettings non optional
